### PR TITLE
Implement a guard for pending user input to avoid deadlocks

### PR DESCRIPTION
**Pull request by echarles**
## References

Fixes https://github.com/jupyterlab/jupyterlab/issues/8632

## Code changes

Add hadPendingInput flag on the OutputArea as guard against multiple cell run while waiting on user input.

## User-facing changes

If you invoke a second time the cell run before submitting your cell input, a message is shown and the cell is not run to avoid the deadlock.

Screenshot 2020-07-24 at 09 52 42

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by echarles** [comment]
The current implementation is a partial fix: If the user invokes another input cell twice while the first is pending, the deadlock still arises. I will now have a look at implementing a guard at the kernel level.

**Comment by echarles** [comment]
This is a screencast with the warning message.

ezgif-4-b01fb4a50536

File: packages/apputils/src/sessioncontext.tsx
Comment (blink1073) [inline packages/apputils/src/sessioncontext.tsx:400]: phas -> has

File: packages/services/src/kernel/kernel.ts
Comment (blink1073) [inline packages/services/src/kernel/kernel.ts:117]: TODO  😄

**Comment by jasongrout** [comment]
Good catch on the deadlock of evaluating the same cell twice and losing the input. That looks really frustrating.

Are we going too far by not allowing *any* cell to be evaluated, though? For example, I can imagine quickly executing every cell in a notebook by just pressing shift-enter a bunch of times. If one of the first cells comes back with an stdin, all of sudden I'm blocked from my fluid workflow. I have to find the stdin, then figure out where it stopped me executing, then continue executing.

Is there a way to *just* block execution of the cell that contains the stdin, as a minimal way to block the deadlock without interrupting fluid user workflows?

**Comment by jasongrout** [comment]
Also, to avoid frustration, is it possible in that dialog to have a button to go to the cell with the stdin?

What if the stdin is over in the console tied to the same kernel. Does it still block execution?

**Comment by echarles** [comment]
> Are we going too far by not allowing any cell to be evaluated, though? For example, I can imagine quickly executing every cell in a notebook by just pressing shift-enter a bunch of times. If one of the first cells comes back with an stdin, all of sudden I'm blocked from my fluid workflow. I have to find the stdin, then figure out where it stopped me executing, then continue executing.

I agree and I also have that kind of flow: request evaluation of a bunch of cells and only after look at results. This is why I have started with a guard on the ouputarea, but then realized that I could ask the execution of another cell with an input without receiving back the answer from the kernel, so without guarding that second cell. Then the door was open to the second invocation resulting in deadlock.

I realize that it is not easy to understand reading this text as even with all the debug tools and log around, it took me time to make sense of it while developing...

Try to think that you get only one `input_request` from the kernel at a time, so you can ask multiple which will cause a deadlock - Kernel can even not be restarted, only terminated works... Not sure why... as I said the real fix would be on the server/kernel? 

So there is a tradeoff: Or we just lock one cell at at time leaving the door open to deadlock, or we lock every cell... which is my second implementation/commit. I have no real strong opinion on those 2 options, but I tend to favor secure behavior over the rest.

> Is there a way to just block execution of the cell that contains the stdin, as a minimal way to block the deadlock without interrupting fluid user workflows?

As I have tried to explain before, you only know that there is a stdin on the kernel response.

> Also, to avoid frustration, is it possible in that dialog to have a button to go to the cell with the stdin?

Let's define the option we want (block all cells or only one), the I can do that (I guess I have the cell reference and can select it for the user).

> What if the stdin is over in the console tied to the same kernel. Does it still block execution?

I have quickly tried with a second notebook connected to the same kernel and have seen that I still had to maybe fix something... but I must also say that the UI behavior for to connect a second notebook the the first notebook kernel gave me surprise (or maybe I was a bit tired...) I was planning to try with a console connected to the same kernel also.

But first let's decide which option to go with: Block only one cell (with deadlock risk) or Block all cells/console connected to the kernel.

**Comment by echarles** [comment]
If we can know in advance that a shell kernel message will generate an `input_request` response, we can have a better solution, but to my understanding there is no way to know that (in that case we would have an easy way to lock the cell

The issue is that once a input_request has been sent, any other one does not come back from the kernel unless you submit the input.

Searching for the `display(...)`string in the cell text is not a valid solution to me.

**Comment by jasongrout** [comment]
In general, I'm pretty strongly in favor of reducing or eliminating interruptions to a user's workflow, so I'd vote for just preventing execution of the cell with the waiting stdin.

**Comment by jasongrout** [comment]
Brainstorming here, another approach is to put the stdin prompt in a modal dialog (perhaps just modal for the document?). That forces a user to respond immediately before doing anything else, but it does lose a bit of context (you don't see the input in line in the cell output).

**Comment by blink1073** [comment]
What if the modal is the whole cell?

**Comment by jasongrout** [comment]
> What if the modal is the whole cell?

That does give extra context, good point. However, that often can be pretty large, especially if you include the input code, whereas the prompt itself is often just a single line.

**Comment by jasongrout** [comment]
In the discussion on #8522 about having a stdin appear when a shell command is executed, we would need the extra context of the output above the prompt to make a decision, so that would indicate we do need at least the output region in a model for stdin input.

**Comment by echarles** [comment]
I love storming my brain.

I would say the modal way is even more flow-breaking....

But maybe we make clear that the requested execution needs user input. I would not be surprised in some notebook flow, that the user does even not realize he needs to submit something.

**Comment by jasongrout** [comment]
> I would say the modal way is even more flow-breaking....

Yes. I bring it up as a better option for the "stop all the things" approach.

**Comment by echarles** [comment]
so a `ModalOutputArea` which can be reused in other use cases (nothing comes to my mind for now, but I am sure some will find ideas for)

**Comment by jasongrout** [comment]
We can already clone the output area (we use it for the "Create New View for Output"). I would imagine we'd just do the same cloning and stick the clone in a modal?

**Comment by echarles** [comment]
> Yes. I bring it up as a better option for the "stop all the things" approach.

Yes, much better. This is how converting an issue into a feature opportunity. I may try something around a `ModalOutputArea` widget this weekend unless I hear the contrary here.

The good think is the guard flag can be removed from the kernel as `the Modal is the Guard`.

**Comment by jasongrout** [comment]
What if multiple clients are connected to the kernel? Does each put up a modal? Is it a document-based modal, so each can put up a modal, but knows when to take it down?

**Comment by echarles** [comment]
> What if multiple clients are connected to the kernel

You mean multiple notebooks and console connected to the same kernel?

> Does each put up a modal?

Only the notebook with the cell that the input would generated the modal...

> Is it a document-based modal, so each can put up a modal, but knows when to take it down?

.... which should be a modal for the complete jlab area to avoid the user running cells in other notebook/console

**Comment by echarles** [comment]
> We can already clone the output area (we use it for the "Create New View for Output"). I would imagine we'd just do the same cloning and stick the clone in a modal?

The clone does not show the currently the input text (which is not a `stream` output..., may be the cause?).

Screenshot 2020-07-24 at 16 45 31

**Comment by mlucool** [comment]
This seems like a step forward. Since like clone, the input box is not restored, open/closing the same notebook still runs into this problem.

The ultimate use case I raised this for was for pdb. The message "Please submit your input" while valid, felt a bit strange here.

**Comment by echarles** [comment]
> Since like clone, the input box is not restored, open/closing the same notebook still runs into this problem.

True, I have forgotten to highlight that this fix does not work if the user refreshes the browser or closes/open the notebook tab.

> The ultimate use case I raised this for was for pdb.

Is there something special to consider for python debugger?

> The message "Please submit your input" while valid, felt a bit strange here.

Although not ideal, would this message be still acceptable for you?

If we find that the client side only fix is not good enough, we need to look at server side. I am not sure if something can be done on the tornado rest/service (which would be valid for all kernels) or if we have to go into the kernel (which would need to be done in IPyKernel, (and/or Xeus?...)).

**Comment by mlucool** [comment]
I verified in the binder that this adds a some guard rails for `pdb`. The typical error mode is an accidental rerun of a cell, which this helps with.

Message proposal:
> Did not run selected cell because there is a cell waiting on input! Run that and try again.

Or something along those lines that moves it from submitting input to waiting on input (the former feels like a form, the latter feels more general).

> If we find that the client side only fix is not good enough, we need to look at server side. 

This seems much better than status quo and its worth at least seeing what people think in practice.

**Comment by echarles** [comment]
I have implemented a modal version of the Stdin component. If the user clicks on cancels the modal (via cancel button, or clicking outside of the box), it is still presented with the input box in the output area.

modal-input

The guard on kernel level is still there and prohibits other cells to be run (I have updated the message based on @mlucool suggestion).

I have tested with a console attached on the same kernel and .... the guard is not triggered... not sure why... but prolly out-of-scope of the reported initial issue.

File: packages/apputils/src/sessioncontext.tsx
Comment (echarles) [reply packages/apputils/src/sessioncontext.tsx:400]: Fixed. Thx

File: packages/services/src/kernel/kernel.ts
Comment (echarles) [reply packages/services/src/kernel/kernel.ts:117]: Done. thx,

**Comment by echarles** [comment]
@blink1073 Based on https://github.com/jupyterlab/jupyterlab/pull/8713#pullrequestreview-454801043https://github.com/jupyterlab/jupyterlab/pull/8713#pullrequestreview-454801043, I have added a test for this.

This is ready for review.

File: packages/outputarea/src/widget.ts
Comment (blink1073) [inline packages/outputarea/src/widget.ts:846]: Missing a docstring

File: packages/outputarea/src/widget.ts
Comment (blink1073) [inline packages/outputarea/src/widget.ts:822]: Missing a docstring

File: packages/services/src/kernel/kernel.ts
Comment (blink1073) [inline packages/services/src/kernel/kernel.ts:117]: TODO comment remaining

**Comment by pepijndevos** [comment]
Eh, would it be possible to just kill the cell if I rerun it while it's waiting.
That's the behaviour I'd expect honestly.
Like... imagine I'm in the debugger and then make a change and rerun the cell.
The current solution would force me to type "exit" in the debugger, rather than just killing the debugger and starting over.

**Comment by echarles** [comment]
@pepijndevos I am not sure if there is a way to trigger the stop of the execution of a particular run request. If there were, there would be anyway not guarantee for this stop request to succeed. This issue does not take into account the debugger, probably a separated issue. could deal with that in a second step.

File: packages/services/src/kernel/kernel.ts
Comment (echarles) [reply packages/services/src/kernel/kernel.ts:117]: Done. Thx

**Comment by echarles** [comment]
Hi, I have rebased on master and taken into account the feedbacks given in the previous comment.

The behavior on the PR is the one shown in https://github.com/jupyterlab/jupyterlab/pull/8713#issuecomment-663853836 (the title of the modal is now ` Waiting on User Input`).

Anything else we should consider here?

**Comment by mlucool** [comment]
Hi @blink1073. Anything blocking merging this?

**Comment by blink1073** [comment]
@echarles, we're still missing one docstring (unresolved comment above), and I think this needs another rebase since several of the checks are failing.

**Comment by goanpeca** [comment]
Ok, making a rebase and will fix the tests and add the missing docstrings. Thanks :)

**Review by blink1073 — APPROVED** [review]
LGTM, thanks!  Let's not merge until we merge #10372 and get the UI test to pass here.

**Comment by goanpeca** [comment]
>LGTM, thanks! Let's not merge until we merge #10372 and get the UI test to pass here.

Sounds good!

**Comment by blink1073** [comment]
Kicking CI

**Comment by goanpeca** [comment]
Should I rebase to get the screenshot updates @blink1073 ?

**Comment by blink1073** [comment]
IIRC because it is a `pull_request` event, it should pick up #10372

**Comment by goanpeca** [comment]
Updated with latest master.

**Comment by blink1073** [comment]
Thanks!

**Comment by choldgraf** [comment]
Hey all - I just discovered that on mybinder.org, we are shipping v 3.1.1 of JupyterLab, and a user has reported that their console is logging `WHAT THE FUCK` when they use the output widget. I did a bunch of digging and arrived at this pull request:

https://github.com/jupyterlab/jupyterlab/commit/33cd7590488455de7f7eea0149498fc24f71c588#diff-4d0f96dfb060e341e538a737407cb92ad9a1003d88037f844a289e11134837abR900-R907

Respectfully, can I just note that I think it is not acceptable for us to put language like this into the source code and especially anything potentially user-facing.

File: packages/outputarea/src/widget.ts
Comment (fperez) [inline packages/outputarea/src/widget.ts:904]: Quick note - this was recently reported to us by a user, and @choldgraf already opened a [repo2docker issue](https://github.com/jupyterhub/repo2docker/issues/1080) to track it and quickly propagate an update to 3.1.11. Unfortunately we happened to update the Binder images to the one that has this code live in it and it means that for a bit, _all_ binder users will see this.

From the logs, it seems this line was introduced on June 13 in [29e50](https://github.com/jupyterlab/jupyterlab/commit/29e50c6225260b0c57c0ab9af68cd873fa8898df#diff-4d0f96dfb060e341e538a737407cb92ad9a1003d88037f844a289e11134837abR904), and it was removed on Augusst 6 in in [ae1846](https://github.com/jupyterlab/jupyterlab/commit/ae1846768e7cb22fbef63cce325fe5ad395b655f#diff-4d0f96dfb060e341e538a737407cb92ad9a1003d88037f844a289e11134837abL919) (thanks @echarles!).

@goanpeca for future reference, it is **not OK**  to put code like this into a submitted PR. I presume this was unintentional and it was left in by accident from internal debugging, but I hope the lesson is that it's never a good idea to leave code like that even in a personal commit, as it's very easy to forget and have it slip by unnoticed for a while. 

When you need log messages that stand out, using things like `*** DBG  - ..... ***` potentially with longer/more visible markers if desired, is both easier to see and doesn't create these issues.

Nothing else to do on this PR since the code is already removed, but I still wanted to leave this note here so that folks keep this in mind and avoid similar incidents in the future. Thanks!

**Comment by blink1073** [comment]
I take responsibility for missing it when reviewing, and also want to thank @goanpeca for all of his hard work and to tell him not to be too hard on himself for an honest mistake.

**Comment by choldgraf** [comment]
Just a note - my goal here isn't to point fingers. I assumed this was an honest mistake, and thought it was important to leave a public record that this wasn't an acceptable thing to have in code. Apologies if my comments came across as combative or accusatory - we are all doing the best we can with the resources we've got.

File: packages/outputarea/src/widget.ts
Comment (goanpeca) [reply packages/outputarea/src/widget.ts:904]: Hi @fperez, yes it was an honest mistake, trying to debug an issue.

I am aware this is not appropriate language, my apologies to the community, this was really an oversight on my side when I made the changes to the original PR.

Note taken

**Comment by goanpeca** [comment]
@choldgraf and @blink1073 no worries, I was just debugging something I did not understand, and the comment was actually venting my frustration at the moment, and I forgot to remove it.

So yes, it was an honest mistake. My apologies to the community, this was really an oversight on my side when I made the changes to the original PR.

And no harm done at all from flagging this issue @choldgraf

**Comment by fperez** [comment]
> So yes, it was an honest mistake. My apologies to the community, this was really an oversight on my side when I made the changes to the original PR.

No worries @goanpeca, and I second Steve's gratitude for the amazing work you're doing across the project, it is impactful and immensely appreciated!

It's important to flag these things mostly so everyone knows what our stance on these matters is across the project, but it is equally important to recognize that we all make mistakes in good faith and that we're all trying to do more than is reasonable with finite resources. So being kind to one another _especially when we make the occasional, very human, small mistake_ is just as important :) Only a project where it's OK to make mistakes can truly dare to explore new horizons and achieve great things, and that is very much the space I hope we build every day.

Thanks again, and keep rocking on the project!

**Comment by echarles** [comment]
Thx all for the work, understanding and openness on this.

> Thanks again, and keep rocking on the project!

+1

**Pull request merged**
